pkgs/build-support/setup-hooks/generate-darwin-launcher.sh

```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env bash
+
+fixupOutputHooks+=('generateDarwinLaunchers $prefix')
+
+# Get a param out of a desktop file. First parameter is the file and the second
+# is a pattern of the key who's value we should fetch.
+getDesktopParam() {
+    local file="$1";
+    local pattern="$2";
+
+    awk -F "=" "/${pattern}/ {print \$2}" "${file}"
+}
+
+# Various icons of different sizes are generated for desktop entries. Here we
+# use the filesize to select the largest icon (which will presumably have the
+# largest resolutioœn) for the launcher.
+getIconFile() {
+    local -r name="$1"
+    local -r candidateIcons=$(find "$out/share/icons/" -name "${name}.png")
+    local biggestSz=0
+    local currentCandidate=""
+
+    for icon in ${candidateIcons}; do
+        local sz=$(stat -c "%s" "${icon}")
+
+        if (( $sz > $biggestSz )); then
+            biggestSz=$sz;
+            currentCandidate="$icon"
+        fi
+    done
+
+    echo "$currentCandidate";
+}
+
+# For a given .desktop file, generate a darwin '.app' launcher for it.
+generateDarwinLauncher() {
+    set -x
+    local -r file="$1"
+    local -r name=$(getDesktopParam "${file}" "^Name")
+    local -r exec=$(getDesktopParam "${file}" "Exec")
+    local -r iconName=$(getDesktopParam "${file}" "Icon")
+    local -r iconFile=$(getIconFile "$iconName")
+
+    mkdir -p "$out/Applications/${name}.app/Contents/MacOS"
+    mkdir -p "$out/Applications/${name}.app/Contents/Resources"
+
+    if [ -n "$iconFile" ]; then
+        png2icns "$out/Applications/${name}.app/Contents/Resources/icon.icns" "$iconFile";
+    fi
+
+    cat > "$out/Applications/${name}.app/Contents/Info.plist" <<EOF
```

Comment:
Can we not reuse make-darwin-application? Granted, there's not much duplication but it seems inelegant anyway unless it's strictly necessary?

Reply:
Agreed, it'd be great if we could reuse that code, I'm just not sure how to call back into Nix from a setup hook?

Reply:
Right, I'm not sure that's possible. Was thinking more of extracting the script parts and then using readFile or whatever is suitable to use them from Nix.

Reply:
Hmm, but wouldn't a call like readFile be invoked when producing the derivation? AFAIK the only code that can be invoked when you have the source for a package available is shell commands? If there's a way to invoke Nix code after the build phase that would do it!

The only other way I can think of removing the duplication is to convert the make-Darwin-application use shell code that is shared.

Reply:
Oh, you're right, it's a setup hook. Having a script that is used by both is kinda what I meant initially though. So make-darwin-application could just call parts of the setup hook or both source or invoke a third script that provides the functions.

Reply:
If I understand correctly, anything we want to reuse in between setup hooks goes to stdenv's setup.sh.

Reply:
Maybe that's a suitable place. Though make-darwin-application isn't a setup hook and sourcing setup.sh from there seems a bit weird.

Reply:
Makes sense. Although, do we want to put something platform specific in stdenv's setup.sh? (I haven't looked but I'm presuming that all platforms use the same setup.sh)

Reply:
Okay, I've managed to get this working. I've create a new package `writeDarwinLauncher` which generates the plist and the executable wrapper.  That is now shared by both functions.